File: wasm/examples/nodejs/javascript/general/derivation.js
const kaspa = require('../../../../nodejs/kaspa');
const {
    Mnemonic,
    XPrv,
    DerivationPath,
    PublicKey,
    NetworkType,
} = kaspa;

kaspa.initConsolePanicHook();

(async () => {

    //const mnemonic = Mnemonic.random();
    const mnemonic = new Mnemonic("hunt bitter praise lift buyer topic crane leopard uniform network inquiry over grain pass match crush marine strike doll relax fortune trumpet sunny silk")
    console.log("mnemonic:", mnemonic);
    const seed = mnemonic.toSeed();
    console.log("seed:", seed);

    // kaspa
    let xPrv = new XPrv(seed);
    // derive full path upto second address of receive wallet
    let pubkey1 = xPrv.derivePath("m/44'/111111'/0'/0/1").toXPub().toPublicKey();
    console.log("publickey", pubkey1.toString())
    console.log("address", pubkey1.toAddress(NetworkType.Mainnet));

    // create receive wallet
    let receiveWalletXPub = xPrv.derivePath("m/44'/111111'/0'/0").toXPub();
    // derive receive wallet for second address
    let pubkey2 = receiveWalletXPub.deriveChild(1, false).toPublicKey();
    console.log("address", pubkey2.toAddress(NetworkType.Mainnet));
    if (pubkey1.toString() != pubkey2.toString()){
        throw new Error("pubkey2 dont match")
    }

    // create change wallet
    let changeWalletXPub = xPrv.derivePath("m/44'/111111'/0'/1").toXPub();
    // derive change wallet for first address
    let pubkey3 = changeWalletXPub.deriveChild(0, false).toPublicKey();
    console.log("change address", pubkey3.toAddress(NetworkType.Mainnet));
    // ---

    //drive address via private key
    let privateKey = xPrv.derivePath("m/44'/111111'/0'/0/1").toPrivateKey();
    console.log("address via private key", privateKey.toAddress(NetworkType.Mainnet))
    console.log("privatekey", privateKey.toString());
    let pubkey4 = privateKey.toPublicKey();
    if (pubkey1.toString() != pubkey4.toString()){
        throw new Error("pubkey4 dont match")
    }

    // xprv with ktrv prefix
    const ktrv = xPrv.intoString("ktrv");
    console.log("ktrv", ktrv)

    //create DerivationPath
    const path = new DerivationPath("m/1'");
    path.push(2, true);
    path.push(3, false);
    console.log(`path: ${path}`);

    // derive by path string
    console.log("xPrv1", xPrv.derivePath("m/1'/2'/3").intoString("xprv"))
    // derive by DerivationPath object
    console.log("xPrv3", xPrv.derivePath(path).intoString("xprv"))
    // create XPrv from ktrvxxx string and derive it
    console.log("xPrv2", XPrv.fromXPrv(ktrv).derivePath("m/1'/2'/3").intoString("xprv"))
    

    // get xpub
    let xPub = xPrv.toXPub();
    // derive xPub
    console.log("xPub", xPub.derivePath("m/1").intoString("xpub"));
    // get publicKey from xPub
    console.log("publicKey", xPub.toPublicKey().toString());
})();

File: python/examples/derivation.py
from kaspa import (
    DerivationPath,
    Mnemonic,
    PublicKey,
    XPrv
)

if __name__ == "__main__":
    mnemonic = Mnemonic("hunt bitter praise lift buyer topic crane leopard uniform network inquiry over grain pass match crush marine strike doll relax fortune trumpet sunny silk")
    seed = mnemonic.to_seed()

    xprv = XPrv(seed)

    # Create receive wallet
    receive_wallet_xpub = xprv.derive_path("m/44'/111111'/0'/0").to_xpub()
    # Derive receive wallet for second address
    pubkey2 = receive_wallet_xpub.derive_child(1, False).to_public_key()
    print(f'Receive Address: {pubkey2.to_address("mainnet").to_string()}')

    # Create change wallet
    change_wallet_xpub = xprv.derive_path("m/44'/111111'/0'/1").to_xpub()
    # Derive change wallet for first address
    pubkey3 = change_wallet_xpub.derive_child(0, False).to_public_key()
    print(f'Change Address: {pubkey3.to_address("mainnet").to_string()}')

    # Derive address via public key
    private_key = xprv.derive_path("m/44'/111111'/0'/0/1").to_private_key()
    print(f'Address via private key: {private_key.to_address("mainnet").to_string()}')
    print(f'Private key: {private_key.to_string()}')

    # XPrv with ktrv prefix
    ktrv = xprv.into_string("ktrv")
    print(f'ktrv prefix: {ktrv}')

    # Create derivation path
    path = DerivationPath("m/1'")
    path.push(2, True)
    path.push(3, False)
    print(f'Derivation Path: {path.to_string()}')

    # Derive by path string
    print(f'{xprv.derive_path("m/1'/2'/3").into_string("xprv")}')
    # Derive by DerivationPath object
    print(f'{xprv.derive_path(path).into_string("xprv")}')
    # Create XPrv from ktrvx string and derive it
    print(f'{XPrv.from_xprv(ktrv).derive_path("m/1'/2'/3").into_string("xprv")}')

    # Get xpub
    xpub = xprv.to_xpub()
    # Derive xpub
    print(xpub.derive_path("m/1").into_string("xpub"))
    # Get public key from xpub
    print(xpub.to_public_key().to_string())


